Rework resources

JaimeBea opened the pull request:
This rework will allow a loading screen to appear in place of waiting for the screen to unfreeze. There are also a lot of fixes related to resources and component loading/saving included within this PR.

Caveats:

- ComponentMeshRenderer's meshID and materialID are no longer accessible from gameplay. Use GetX and ChangeX instead.
- Calling ChangeScene will make the scene change immediately as soon as it gets loaded. If you want to preload it in order to change scene later PreloadScene will do the trick.
- Load methods on components are called in a separate thread, so reference counts need to be updated outside this function.
- Resources now have two different loading functions, Load and FinishLoading. The first one will be called in a different thread, so it has to be used carefully. FinishLoading will be executed after loading in the main thread and is safer to use.
- GameObject's Init and Start functions have been changed. Init will be now called in the main thread after loading and Start will be called when starting the game.



EDIT:
**Use the branch "feature/rework-resources-test" to test the changes with the Assets included.**

To facilitate the division of the code review, check here the classes that you have reviewed, so the others can avoid them if they want, and we will be able to track if everything has been reviewed.
It would be nice to have three checks (:heavy_check_mark::heavy_check_mark::heavy_check_mark:) at each file, a minimum of two (:heavy_check_mark::heavy_check_mark:)

- AnimationController
- COMPONENTS
  - ComponentAgent
  - ComponentAnimation
  - ComponentAudioSource
  - ComponentBillobard
  - ComponentLight :heavy_check_mark:
  - ComponentMeshRenderer
  - ComponentObstacle
  - ComponentParticleSystem
  - ComponentScript
  - ComponentSkybox
  - ComponentTrail
  - ComponentTransform :heavy_check_mark:
  - PHYSICS
    - ComponentBoxCollider :heavy_check_mark:
    - ComponentCapsuleCollider :heavy_check_mark:
    - ComponentSphereCollider :heavy_check_mark:
  - UI
    - ComponentEventSystem
    - ComponentImage
    - ComponentVideo :heavy_check_mark:
 - GameObject :heavy_check_mark:
 - GameplaySystems
 - IMPORTERS
   - AudioImporter
   - ClipImporter
   - FontImporter
   - MaterialImporter :heavy_check_mark:
   - ModelImporter
   - NavMeshImporter :heavy_check_mark:
   - PrefabImporter :heavy_check_mark:
   - SceneImporter
   - ScriptImporter :heavy_check_mark:
   - ShaderImporter :heavy_check_mark:
   - SkyboxImporter :heavy_check_mark:
   - StateMachineGenerator
   - StateMachineImporter :heavy_check_mark:
   - TextureImporter :heavy_check_mark:
   - VideoImporter :heavy_check_mark:
 - LightFrustum :heavy_check_mark:
 - MODULES
   - ModuleCamera :heavy_check_mark:
   - ModuleEditor :heavy_check_mark:
   - ModuleNavigation
   - ModulePhysics :heavy_check_mark:
   - ModuleProject
   - ModuleRender
   - ModuleResources
   - ModuleScene
   - ModuleTime :heavy_check_mark:
   - ModuleUserInterface
 - NavMesh
 - Panels (in general)  :heavy_check_mark:
 - RESOURCES
   - Resource  :heavy_check_mark:
   - ResourceAnimation
   - ResourceAudioClip
   - ResourceClip
   - ResourceFont
   - ResourceMaterial
   - ResourceMesh
   - ResourceNavMesh
   - ResourcePrefab
   - ResourceScene
   - ResourceShader
   - ResourceSkybox
   - ResourceStateMachine
   - ResourceTexture
 - Scene
 - StateMachineManager
 - TesseractEvent
 - UTILS
   - FileDialog
   - ImGuiUtils
   - PathUtils
   - Trail

FrancPS commented on Project/Source/AnimationController.cpp:
```diff
@@ -13,38 +13,32 @@
-	int intPart = (int) currentSample;
-
-	return intPart;
+	unsigned currentSamble = clip.beginIndex + static_cast<unsigned>((currentTime * (clip.keyFramesSize)) / clip.duration);
```
curretSamPle

FrancPS reviewed (approved):
Accepted code for Video and Physics.

Pending QA testing.

JaimeBea commented:
> Bugs:
> 
> * Engine crashes if you load the scene (after compile all the scripts) with an object that has a ComponentScript without any script assigned.
> * Engine crashes if you compile the scripts having a scene loaded with an object that has a ComponentScript without any script assigned.
> * Engine crashes if you spam the shoot button while the video is playing. This bug doesn't occur in Develop or Penteract Repository.
> * Some enemies state machine bugged.
> * Engine always crashed on stop, DestroyGameObject (gameObject) the gameObject is garbage
> * Play button is active while loading a scene, if you clicked during the load, the engine crashes
> * In engine You can load a scene while playing, then if you stop the game, and click play button again, the engine crashes
> * Memory leaks when loading level 2 and exit
> * Cannot play in DEBUG less than 10 FPS
> * Prefabs don't work
> * Component Billboard texture don't show

* Fixed.
* Fixed.
* Fixed.
* Needs more testing.
* Fixed.
* Will fix in another PR. Avoid doing this meanwhile.
* Will fix in another PR. Avoid doing this meanwhile.
* Will fix in another PR.
* Will fix in another PR.
* Fixed.
* Will fix in another PR. Happened in develop too.

Mefiso commented on Project/Source/Components/ComponentTransform.cpp:
```diff
@@ -226,10 +207,15 @@ float3 ComponentTransform::GetGlobalPosition() {
-	newTransform_.Orthogonalize3();
-	newTransform_.Orthonormalize3();
-	return Quat(newTransform_.SubMatrix(3, 3));
+	float3x3 globalRotation = globalMatrix.RotatePart();
```
I believe there is a problem here in this method. If an ancestor of a rotated GameObject has a non-uniform scale, the global rotation matrix may not be orthogonal, and although you are normalizing the axes, that does not reorthogonalize the 3x3 matrix. Then, the Quat will not be correctly constructed, leading to an incorrect global rotation (I think it will log a failed assumption).

Maybe I'm wrong, someone should test it. I can test it on sunday, but not before.

lowysole replied:
The global matrix is orthogonalized in the CalculateGlobalMatrix().
`globalMatrix.Orthogonalize3(); // Solution for non-uniform scaled objects`
But I'm testing all this part in detail

magalenyo commented on Project/Source/Importers/ModelImporter.cpp:
```diff
@@ -461,8 +460,8 @@ static void ImportNode(const char* modelFilePath, JsonValue jMeta, const aiScene
-			meshRenderer->meshId = ImportMesh(modelFilePath, jMeta, assimpMesh, resourceIndex, bones);
-			meshRenderer->materialId = materialIds[assimpMesh->mMaterialIndex];
+			meshRenderer->SetMeshID(ImportMesh(modelFilePath, jMeta, assimpMesh, resourceIndex, bones));
+			meshRenderer->SetMaterialID(materialIds[assimpMesh->mMaterialIndex]);
```
Shouldn't that be using the ChangeMaterial instead of SetMaterialID? If not, where are you increasing the reference?

JaimeBea replied:
This is on purpose. The reference gets increased when initializing the component.

magalenyo replied:
OK, ty

magalenyo commented on Project/Source/Components/ComponentScript.cpp:
```diff
@@ -24,8 +24,34 @@
+}
+
 void ComponentScript::Start() {
-	if (scriptInstance && App->time->HasGameStarted() && App->scene->scene->sceneLoaded) {
```
Why have these lines been deleted? What if a Script needs the reference of another go that hasn't yet been initialized? Are you calling the Script->Start() from somewhere else?

JaimeBea replied:
Start only gets called when this condition is true, so it was redundant.

magalenyo commented on Project/Source/GameObject.cpp:
```diff
@@ -170,12 +174,6 @@ void GameObject::AddMask(MaskType mask_) {
 	case MaskType::PLAYER:
 		mask.bitMask |= static_cast<int>(mask_);
 		break;
-	case MaskType::CAST_SHADOWS:
```
Is this deprecated code that is being removed?

JaimeBea replied:
Yes, this was deprecated when merging shadows with Jordi.

magalenyo commented on Project/Source/Modules/ModuleScene.cpp:
```diff
@@ -94,20 +97,14 @@ bool ModuleScene::Start() {
 	App->files->CreateFolder(NAVMESH_PATH);
 #endif
 
+	App->scene->CreateEmptyScene();
```
I'm assuming this is necessary for the Resources Rework? Otherwise, it's an unnecessary call outside the else statement you've deleted this from.

JaimeBea replied:
Yeah, it is important that the scene is never null.

magalenyo reviewed (changes requested):
Hi,

First of all, thanks for the huge effort you have made doing this rework. This is truely remarkable. 😄 

Now, I have found several important bugs I will be describing down below. There are many modules or componentes I haven't tested because I simply do not know how to or just didn't think of it. There are probably other unknown bugs that we are currently not seeing because we are not working fully with this rework, but that we are going to be facing when we are building new scenes.

This review is divided in the following parts: Engine on Debug Mode, Engine on Release Mode, Game Mode and a Summary.

### Engine on Debug

1. Debug mode extremely laggy (15 fps peak), this is kinda a legacy problem, but still exists.
2. It took almost 2.5 minutes to fully load+compile the scene.

Stopped testing Debug because it was untestable.

### Engine on Release

1. Materials do not change correctly (tested on already existing material, dissolve material doors after defeating enemies on Level1). On Level1 spawn, the door material is transparent, after defeating the enemies and moving forward, when door closes up, the material has an Opaque material.
2. On hit effects particles are not working correctly and particles seems to be off (something about colors / extra effects). Also, subemitters seem to be broken since Fang's ultimate (the most evident one) doesn't have the hit effect on enemies.
3. Doesn't fade in from Level1 to Level2 (probably due to Scene loading).
4. In an empty scene, after creating a material, setting it to an object and modifying it, after saving material, the engine crashes: 
![image](https://user-images.githubusercontent.com/27434068/134788166-e511118e-c6c5-401b-b1dd-94fbf5e95094.png)
![image](https://user-images.githubusercontent.com/27434068/134788169-c85d0212-975b-4ff4-8743-e6b2a5ea0cff.png)

5. When spawning prefabs on Level1 test, this happened:
![image](https://user-images.githubusercontent.com/27434068/134788175-c6b6acd1-f409-4083-a886-801fb2952a80.png)

6. Level1 initially spawned enemies are not correctly spawned, they are frozen:
![image](https://user-images.githubusercontent.com/27434068/134788182-12d13744-4f89-4f06-832d-2b9424e2b4b1.png)

7. On Level1 test, UI got bugged and didn’t swap canvas, also, the switch sound is bugged and played more than once. (the sound might be caused by Gameplay, but the UI bug only got it reproduced in this PR).
![image](https://user-images.githubusercontent.com/27434068/134788199-cd70916d-7e59-484c-bbd1-a85ee5caaf94.png)

8. When loading one scene to another, everything keeps playing until the new scene is loaded: (the gif is way too big so I'm pasting the URL). It can be easily reproduced spawning Ranged enemies, pressing Pause Menu, and pressing "Return to Menu".
https://gyazo.com/736ec208e176e9390475a53e4d4cfe1c

9. When loading one scene to another (from Level1 to StartScene from Menu), scene code broke up since it’s still accessing GameObjects that may already be deleted:
![image](https://user-images.githubusercontent.com/27434068/134788253-ce821113-c8eb-4d2c-bc96-80311ebaf703.png)


10. When following this next flow on Play: Level 1, open Main Menu and press "Return to Menu", on StartScene press "Start": Fang appears in the elevator and UI is displaced (this bug was caused by retrying the game a couple of times): 
![image](https://user-images.githubusercontent.com/27434068/134788256-d4000f8b-61a5-4e5e-9741-422f0e45242f.png)

11. I was in Pause mode, then pressed Play, and then Stop, and then this happened:
![image](https://user-images.githubusercontent.com/27434068/134788305-1e1b08ba-86d5-4e17-9df0-fbfdd1b02b50.png)


12. Navigation Panel Editor is empty in a new Scene. This is caused because you are forcing the user to have a NavMesh to show the NavMesh Building tab, which makes no sense.
![image](https://user-images.githubusercontent.com/27434068/134788377-88d2e96b-c6d9-4df6-9c6b-8c4b5e5e8e8d.png)
![image](https://user-images.githubusercontent.com/27434068/134788385-d65b524f-dd12-48a1-a13b-2c348e6b6f2c.png)


13. Resource Tab is empty even though the ID isn’t empty or null. This is caused because the Resource returned is null.
![image](https://user-images.githubusercontent.com/27434068/134788402-8939ac7f-9843-449a-8535-10f24d207cf9.png)
![image](https://user-images.githubusercontent.com/27434068/134788403-fe2c61ef-55de-4430-88f6-73919352b378.png)


14. From a scene with a camera with Frustum Calling Active (for example, an empty scene), and loading a new scene (Level 1), the engine crashed in the next lines:
![image](https://user-images.githubusercontent.com/27434068/134788450-b81b6332-e6d7-4c7b-b2f1-87a959f51687.png)
![image](https://user-images.githubusercontent.com/27434068/134788454-ad8c4795-c297-40e7-bb5a-472f76887cbc.png)


15. When Adding a Video Gameobject on the scene, loading a Resource and pressing Play from ComponentVideo’s OnEditorUpdate, the engine crashes.
![image](https://user-images.githubusercontent.com/27434068/134788459-7b039ee6-4b8a-4e51-bc8f-e70fe4101e9e.png)


16. When Adding a Video GameObject on the scene, loading a Resource and pressing Stop from ComponentVideo’s OnEditorUpdate, the engine crashes.
![image](https://user-images.githubusercontent.com/27434068/134788466-f7ba0fc3-89ac-4b54-a608-c7f92344f3e4.png)


### Game mode

1. In game mode, it loads the grid (default scene), then tries to load the scene but it never loads it (there’s only a black screen with the level music).

### Summary
- Debug Engine was untestable.
- Release Engine: this PR introduces many bugs that need reviewing, and many others that I couldn't test because I don't know how to, for example animations. Also, important features like NavMesh building are currently unavailable.
- Game was untestable.

**I strongly advice NOT TO merge this PR and grab all the "fixes" necessary and add them in a different PR**, for example release memory functions, or the additional code to correctly increase the number of references of a Resource, for example the ChangeMaterial and ChangeMesh functions of the ComponentMeshRenderer.

JaimeBea commented:
Engine on Debug
- These problems are also in the main branch

Engine on Release
- Materials do not change correctly: Investigating.

- Particle effect bugs: Fixed.

- No fade in on level 2: Investigating.

- Material save crash: Can't reproduce.

- Prefab spawn crash: Investigating.

- Frozen enemies: This is related to the animation bugs. Working on it.

- UI switch bug: Investigating.

- Playing while returning to menu: Investigating.

- GetCurrentState crash: This is a known problem. It's related to the animation bugs, I'm working with Saul and Joan Marc to fix them.

- Fang in elevator: This is a gameplay bug that didn't happen before because checkpoints didn't work. Now they work and it's making Fang appear there.

- GameCamera crash: Fixed.

- Navigation: This problem is also in the main branch. It needs fixing.

- Empty Resource Tab: Working on it.

- Frustum culling camera crash: Fixed.

- Video component crash: Investigating.

Game mode
- I can't reproduce the error.
